agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/SpanProcessor.java

```diff
@@ -0,0 +1,150 @@
+        IncludeExclude normalizedExclude = config.exclude != null ? getNormalizedIncludeExclude(config.exclude) : null;
+        List<String> fromAttributes = new ArrayList<>();
+        if (config.name.fromAttributes != null) {
+            fromAttributes.addAll(config.name.fromAttributes);
+        }
+        List<String> toAttributeRules = new ArrayList<>();
+        if (config.name.toAttributes != null) {
+            toAttributeRules.addAll(config.name.toAttributes.rules);
+        }
+        List<Pattern> toAttributeRulePatterns = new ArrayList<>();
+        if (config.name.toAttributes != null) {
+            for (String rule : config.name.toAttributes.rules) {
+                toAttributeRulePatterns.add(Pattern.compile(rule));
+            }
+        }
+        List<List<String>> groupNames = getGroupNames(toAttributeRules);
+        String separator = config.name.separator != null ? config.name.separator : "";
+        return new SpanProcessor(normalizedInclude, normalizedExclude, true,
+                fromAttributes, toAttributeRules, toAttributeRulePatterns, groupNames, separator);
+    }
+
+    private static List<List<String>> getGroupNames(List<String> toAttributeRules) {
+        List<List<String>> groupNames = new ArrayList<>();
+        for (String rule : toAttributeRules) {
+            List<String> subGroupList = new ArrayList<>();
+            Matcher matcher = capturingGroupNames.matcher(rule);
+            while (matcher.find()) {
+                subGroupList.add(matcher.group(1));
+            }
+            groupNames.add(subGroupList);
+        }
+        return groupNames;
+    }
+
+    //fromAttributes represents the attribute keys to pull the values from to generate the new span name.
+    public SpanData processFromAttributes(SpanData span) {
+        if (spanHasAllFromAttributeKeys(span)) {
+            StringBuffer updatedSpanBuffer = new StringBuffer();
+            ReadableAttributes existingSpanAttributes = span.getAttributes();
+            for (String attributeKey : fromAttributes) {
+                updatedSpanBuffer.append(existingSpanAttributes.get(AttributeKey.stringKey(attributeKey)));
+                updatedSpanBuffer.append(separator);
+            }
+            // Removing the last appended separator
+            if (separator.length() > 0) {
+                updatedSpanBuffer.setLength(updatedSpanBuffer.length() - separator.length());
+            }
+            return new MySpanData(span, span.getAttributes(), new String(updatedSpanBuffer));
+        }
+        return span;
+    }
+
+    private boolean spanHasAllFromAttributeKeys(SpanData span) {
+        if (fromAttributes.isEmpty()) return false;
+        ReadableAttributes existingSpanAttributes = span.getAttributes();
+        for (String attributeKey : fromAttributes) {
+            if (existingSpanAttributes.get(AttributeKey.stringKey(attributeKey)) == null) return false;
```

Comment:
can fromAttributes be List<AttributeKey<?>> to avoid creating AttributeKey each time?